src/traits.jl

```diff
@@ -121,16 +114,36 @@ For example,
     color_type(RGB{Float32}) == RGB{Float32}
     color_type(ARGB{N0f8})     == RGB{N0f8}
 """
+color_type(::Type{TransparentColor})        = Color
 color_type{C<:Color}(::Type{C}) = C
+color_type(c::Colorant) = color_type(typeof(c))
+
+# Return the number of components in the color
+# Note this is different from div(sizeof(c), sizeof(eltype(c))) (e.g., RGB1)
+length(c::Colorant) = length(typeof(c))
+
+if isdefined(Core, :UnionAll)
+include_string("""
+    length(::Type{C}) where C<:(Colorant{T,N} where T) where N = N
+    # This definition should be unnecessary, but julia currently incorrectly
+    # dispatches to the first definition below, even when the second is
+    # applicable.
+    _color_type(::Type{TC}) where TC<:(TransparentColor{C, T, N} where T where N) where C = C
+    color_type(::Type{TC}) where TC<:TransparentColor =
+      isa(TC, UnionAll) ? parameter_upper_bound(TC, 1) : _color_type(TC)
+    color_type(::Type{TC}) where TC<:(TransparentColor{C, T, N} where T where N) where C = C
+""")
+else
+    length{T,N}(::Type{Colorant{T,N}}) = N
+    @pure length{C<:Colorant}(::Type{C}) = length(supertype(C))
```

Comment:
...nassty old typesystem don't like it when hobbitses reorder these methods...

```julia
julia> length(RGB)
ERROR: MethodError: no method matching length(::Type{Any})
Closest candidates are:
  length(::SimpleVector) at essentials.jl:168
  length(::Base.MethodList) at reflection.jl:256
  length(::MethodTable) at reflection.jl:322
  ...
 in length(::Type{ColorTypes.RGB}) at /home/tim/.julia/v0.6/ColorTypes/src/traits.jl:138
```
This package was partially an art in catering to the specific bugs in the older type system.